src/hardware/adlib_gold.cpp

```diff
@@ -0,0 +1,388 @@
+	for (auto &f : lowshelf)
+		f.setup(sample_rate, cutoff_freq, gain_db, slope);
+}
+
+void StereoProcessor::SetHighShelfGain(const double gain_db) noexcept
+{
+	DEBUG_LOG_MSG("ADLIBGOLD.STEREO: SetLowShelfGain: %fdB", gain_db);
+
+	constexpr auto cutoff_freq = 3500.0;
+	constexpr auto slope       = 0.4;
+	for (auto &f : highshelf)
+		f.setup(sample_rate, cutoff_freq, gain_db, slope);
+}
+
+StereoProcessor::~StereoProcessor() {}
+
+constexpr auto volume_value_0db   = 60;
+constexpr auto shelf_eq_value_0db = 6;
+
+void StereoProcessor::Reset() noexcept
+{
+	ControlWrite(StereoProcessorControlReg::VolumeLeft, volume_value_0db);
+	ControlWrite(StereoProcessorControlReg::VolumeRight, volume_value_0db);
+	ControlWrite(StereoProcessorControlReg::Bass, shelf_eq_value_0db);
+	ControlWrite(StereoProcessorControlReg::Treble, shelf_eq_value_0db);
+
+	StereoProcessorSwitchFunctions sf = {};
+	sf.source_selector                = static_cast<uint8_t>(
+                StereoProcessorSourceSelector::Stereo1);
+	sf.stereo_mode = static_cast<uint8_t>(StereoProcessorStereoMode::LinearStereo);
+
+	ControlWrite(StereoProcessorControlReg::SwitchFunctions, sf.data);
+}
+
+void StereoProcessor::ControlWrite(const StereoProcessorControlReg reg,
+                                   const uint8_t data) noexcept
+{
+	auto calc_gain = [](const uint8_t value) {
+		constexpr auto VolumeControlBits = 6;
+		constexpr auto VolumeControlMask = (1 << VolumeControlBits) - 1;
+
+		constexpr auto max_gain_db = 6.0;
+		constexpr auto min_gain_db = -128.0;
+		constexpr auto step_db     = 2.0;
+
+		auto val = static_cast<int8_t>(value & VolumeControlMask) -
+		           volume_value_0db;
+		auto gain_db = clamp(val * step_db, min_gain_db, max_gain_db);
+		return decibel_to_gain(gain_db);
+	};
+
+	auto calc_filter_gain_db = [](const uint8_t value) {
+		constexpr auto FilterControlBits = 4;
+		constexpr auto FilterControlMask = (1 << FilterControlBits) - 1;
+
+		constexpr auto max_gain_db = 15.0;
+		constexpr auto min_gain_db = -12.0;
+		constexpr auto step_db     = 3.0;
+
+		auto val = static_cast<int8_t>(value & FilterControlMask) -
+		           shelf_eq_value_0db;
+		return clamp(val * step_db, min_gain_db, max_gain_db);
+	};
+
+	switch (reg) {
+	case StereoProcessorControlReg::VolumeLeft:
+		gain.left = static_cast<float>(calc_gain(data));
+		break;
+
+	case StereoProcessorControlReg::VolumeRight:
+		gain.right = static_cast<float>(calc_gain(data));
+		break;
+
+	case StereoProcessorControlReg::Bass:
+		SetLowShelfGain(calc_filter_gain_db(data));
+		break;
+
+	case StereoProcessorControlReg::Treble:
+		SetHighShelfGain(calc_filter_gain_db(data + 1));
+		break;
+
+	case StereoProcessorControlReg::SwitchFunctions:
+		auto sf = StereoProcessorSwitchFunctions{data};
+
+		source_selector = StereoProcessorSourceSelector(
+		        static_cast<uint8_t>(sf.source_selector));
+		stereo_mode = StereoProcessorStereoMode(
+		        static_cast<uint8_t>(sf.stereo_mode));
+		break;
```

Comment:
This `case` will need a scope because we're making a new temporary inside of it.
(very annoying & subtle aspect of switches that tripped me up, https://stackoverflow.com/questions/92396/why-cant-variables-be-declared-in-a-switch-statement)

Reply:
Yeah I agree. It tripped me up too a few times, I'm used to pattern matching blocks in Scala 😄